2016/proposals/pravendra-singh.md

```diff
@@ -0,0 +1,220 @@
+
+## Technical Details
+
+The application will be powered by Django, as we also want it to work with AMY. It will use Plotly.js for visualizing different response types, which is a open source [project](https://plot.ly/javascript/open-source-announcement/).
+
+Going by the "*Object-oriented analysis and design*" approach for system design, we can categorize the application in the following sections.
+
+###### 1. Visual representation of the survey questions
+
+Both the type of surveys have multiple questions of varying classes. Ideally, a good visualizer should respect the inherent attributes of the dataset. That's why the application will employ suitable visualization for different types of questions.
+
+According to this [explanation](https://github.com/numfocus/gsoc/issues/77#issuecomment-192500549), here is a list of suitable charts.
+
+| Question Type  | Suitable Chart Type | Sample Question |
+| ------------- | ------------- | ------------- |
+| Multiple choice - Single answer  | Bar and Pie chart  | When are you taking this survey? |
+| Multiple choice - Multiple answer  | Chord diagram  |  What is your domain of research/study? |
+| Yes-No-Maybe type  | Horizontal Stacked Bar chart  |  How would you describe your ability to do the following tasks? |
+| Matrix type  | Heatmap Matrix chart  | How would you rate your motivation to learn about these topics? |
+| Text as an input  | Text-Cloud OR Table OR Semantic Analysis | In a few words, what is your most important reason for attending this workshop? |
+
+###### 2. Response-Cache service
+
+As we are already aware of the order of total responses we get from our attendees, having a "*Data Cache Layer*" is an intelligent solution. Rate Limitations exposed by SurveyMonkey API is another reason behind implementing a cache service.
+
+The *Response-Cache Service* will be responsible for serving cached responses, and hence plots. It will leverage the *offline* mode of Plotly so that the caching service is only dependent on the API responses and idempotent in terms of generated plots.
+
+The service will use *Heartbeat Mechanism* to invalidate cache content with respect to the API response periodically. Workshop identifiers (ID) will be used to collect and invalidate responses for a particular event.
+
+**Interaction between SurveyMonkey API and Response-Cache Service**
+```
+           +-------------------+        
+           |    SurveyMonkey   |        
+           |         API       |        
+           +--^-------------^--+        
+              |  Heartbeat  |           
+           +--v-------------v--+        
+           |   Response Cache  |        
+           |      Service      |        
+           +-------------------+
+```
+
+This service consists of two elements "Cache Storage" [CS] and "Permanent Storage" [PS], using "Redis" and "RethinkDB" respectively.
```

Comment:
@pbanaszkiewicz,

I'm writing out the architecture about the "Data Aggregation Service" **DAS** here, please correct me if you find something fishy.

SurveyMonkey API provides an elegant solution in the form of [Polling](https://developer.surveymonkey.com/docs/guides/polling-guide/), roughly similar to the [Webhooks](https://en.wikipedia.org/wiki/Webhook).

Using this API feature, we can find out newly arrived responses to a survey, and we don't need to request all of the survey response data when searching for new responses.

Our service can remember the last time if fetched the API endpoints in the form of a variable, and use this time-string to find out only those responses that arrived after this time-string.

There will be a **Analyser Service** to properly format the response data received from the SurveyMonkey API and feed it to the database. It will use the details as **"workshop_id"** and the answer to the question **"what's your workshop?"** to format the data.

Initially, when we are in a situation of [Cold Start](https://en.wikipedia.org/wiki/Cold_start) and don't have any survey responses as well as the timestamp of last data-fetch operation, we can collect all the data. But this will be a one-time operation as the polling will take care of the new responses after it.

If I break it to the module level, here is a coupling diagram for the service and its components.

```
+-------------------------------------+
|           SurveyMonkey API          |
+-----+-------------------------+-----+
      |                         |
      |                         |
      |                         |
      |                         |
+-------------------------------------+
|     |                         |     |
| +---v-----+            +------v---+ |
| | Starter |            | Listener | |
| +---+-----+            +------+---+ |
|     |                         |     |
+-------------------------------------+
      |                         |
      |                         |
+-------------------------------------+
|     v                         v     |
|          Analyser Service           |
|                  +                  |
+-------------------------------------+
                   |
                   |
                   |
+-------------------------------------+
|                  v                  |
|              Database               |
|                                     |
+-------------------------------------+
      [Data Aggregation Service]
```

Description of individual modules:
* **Starter** - Collects all the survey responses once at the time of application starting.
* **Listener** - Periodically (like Heartbeat, as I had said), this module will look for new survey responses using the Polling feature provided by SurveyMonkey API.
* **Analyser Service** - This module will format the response data using "workshop_id" and the name of the workshop.

I have already responded your question to why I want to use RethinkDB, and also explained the details about caching.

Please correct me if I'm wrong somewhere.